File: C++/Comments/Backtracking/22. Generate Parentheses.cpp
class Solution {
        
public:
    void backtrack(string s, int open, int close, int n, vector<string>& res) {
        if (s.size() == 2*n && open == n && close == n) {
            res.emplace_back(s);
        }
        if (open+1 <= n) {
            backtrack(s+"(", open+1, close, n, res);
        }
        if (close+1 <= n && open>close){
            backtrack(s+")", open, close+1, n, res);
        }
    }
    
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        backtrack("", 0, 0, n, res);
        return res;
    }
};

File: Python/22. Generate Parentheses.py
from typing import List

class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        def helper(s, open_, close_):
            if len(s) == 2*n and open_ == n and close_ == n:
                res.append(s)
            if open_+1 <= n:
                helper(s+'(', open_+1, close_)
            if close_+1 <= n and open_ > close_:
                helper(s+')', open_, close_+1)
        helper('', 0, 0)
        return res